core/src/main/java/zipkin/reporter/internal/BoundedQueue.java

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2016 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin.reporter.internal;
+
+import com.google.auto.value.AutoValue;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import zipkin.internal.Nullable;
+
+/** Special queue that can measure the size of entries as they are added. */
+public final class BoundedQueue<E> {
+
+  public interface ByteCounter<E> {
+    /** Returns the count of bytes in the element. */
+    int byteCount(E element);
+  }
+
+  final int maxBatchSize;
+
+  final ReentrantLock lock = new ReentrantLock(false);
+  final Condition available = lock.newCondition();
+
+  @Nullable final ByteCounter<E> byteCounter;
+  final int maxByteCount; // only honored when byteCounter is set
+
+  final Object[] items;
+  int count;
+  int writeIndex;
+  int readIndex;
+
+  public static <E> BoundedQueue<E> create(int size, int maxBatchSize) {
+    return new BoundedQueue<>(size, maxBatchSize, null, Integer.MAX_VALUE);
+  }
+
+  public static <E> BoundedQueue<E> create(int size, ByteCounter<E> byteCounter, int maxByteCount) {
+    return new BoundedQueue<>(size, Integer.MAX_VALUE, byteCounter, maxByteCount);
+  }
+
+  @AutoValue
+  static abstract class ElementWithByteCount<E> {
+    static <E> ElementWithByteCount<E> create(E element, int byteCount) {
+      return new AutoValue_BoundedQueue_ElementWithByteCount<>(element, byteCount);
+    }
+
+    abstract E element();
+
+    abstract int byteCount();
+  }
+
+  BoundedQueue(int size, int maxBatchSize, ByteCounter<E> byteCounter, int maxByteCount) {
+    this.items = new Object[size]; // cannot instantiate a generic type
+    this.maxBatchSize = maxBatchSize;
+    this.byteCounter = byteCounter;
+    this.maxByteCount = maxByteCount;
+  }
+
+  /** Returns true if the element could be added or false if the queue is full */
+  public boolean offer(E element) {
+    Object toAdd = element;
+    if (byteCounter != null) {
+      int byteCount = byteCounter.byteCount(element);
```

Comment:
nope. this just reports the size of the encoded buffer. ex in the case of a
byte array bytes.length

On Tue, Aug 23, 2016 at 11:14 PM, Felix <notifications@github.com> wrote:

> In core/src/main/java/zipkin/reporter/internal/BoundedQueue.java
> <https://github.com/openzipkin/zipkin-reporter-java/pull/7#discussion_r75887455>
> :
>
> > +
> > +    abstract int byteCount();
> > +  }
> > +
> > +  BoundedQueue(int size, int maxBatchSize, ByteCounter<E> byteCounter, int maxByteCount) {
> > +    this.items = new Object[size]; // cannot instantiate a generic type
> > +    this.maxBatchSize = maxBatchSize;
> > +    this.byteCounter = byteCounter;
> > +    this.maxByteCount = maxByteCount;
> > +  }
> > +
> > +  /** Returns true if the element could be added or false if the queue is full */
> > +  public boolean offer(E element) {
> > +    Object toAdd = element;
> > +    if (byteCounter != null) {
> > +      int byteCount = byteCounter.byteCount(element);
>
> Does that mean spans have to be serialized twice? Once to get the size and
> once more to actually send them over the wire?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin-reporter-java/pull/7/files/a4fd2cf9406f13fe83d31a3fce7603fff0e39fb6#r75887455>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD619jt1hGsGJzLyEmav31qI5CGXO8xks5qiw5EgaJpZM4JpQw9>
> .
>